File: src/solutions/java/N917ReverseOnlyLetters.java
package solutions.java;

public class N917ReverseOnlyLetters {

    public static String reverseOnlyLetters(String input) {
        char[] chars = input.toCharArray();
        int i = 0;
        int j = chars.length - 1;
        while (i < j) {
            if (isLetter(chars[i]) && isLetter(chars[j])) {
                char temp = chars[i];
                chars[i] = chars[j];
                chars[j] = temp;
                i++;
                j--;
            }
            if (!isLetter(chars[i])) {
                i++;
            }
            if (!isLetter(chars[j])) {
                j--;
            }
        }
        return String.valueOf(chars);
    }

    private static boolean isLetter(char input) {
        return 'a' <= input && input <= 'z' || 'A' <= input && input <= 'Z';
    }

    public static void main(String[] args) {
        assert reverseOnlyLetters("ab-cd").equals("dc-ba");
        assert reverseOnlyLetters("a-bC-dEf-ghIj").equals("j-Ih-gfE-dCba");
        assert reverseOnlyLetters("Test1ng-Leet=code-Q!").equals("Qedo1ct-eeLg=ntse-T!");
    }
}

File: src/solutions/scala/N917ReverseOnlyLetters.scala
package solutions.scala

import scala.annotation.tailrec

object N917ReverseOnlyLetters {

  def reverseOnlyLetters(input: String): String = {
    @tailrec
    def reverseOnlyLetters(arr: Array[Char], i: Int, j: Int): Array[Char] = {
      if (i >= j) arr
      else (arr(i).isLetter, arr(j).isLetter) match {
        case (true, true) =>
          val temp = arr(i)
          arr(i) = arr(j)
          arr(j) = temp
          reverseOnlyLetters(arr, i + 1, j - 1)
        case (false, true) => reverseOnlyLetters(arr, i + 1, j)
        case (true, false) => reverseOnlyLetters(arr, i, j - 1)
        case _ => reverseOnlyLetters(arr, i + 1, j - 1)
      }
    }

    reverseOnlyLetters(input.toCharArray, 0, input.length - 1).mkString
  }

  def main(args: Array[String]): Unit = {
    assert(reverseOnlyLetters("ab-cd") == "dc-ba")
    assert(reverseOnlyLetters("a-bC-dEf-ghIj") == "j-Ih-gfE-dCba")
    assert(reverseOnlyLetters("Test1ng-Leet=code-Q!") == "Qedo1ct-eeLg=ntse-T!")
  }
}
